lerobot/common/policies/octo/modeling_octo.py

```diff
@@ -0,0 +1,650 @@
+
+        if add_positional_encoding:
+            pos_emb_max_len = n_obs_tokens + n_readouts * n_obs
+            self.pos_emb = AddPositionalEncoding(pos_emb_max_len, embed_dim, dropout=dropout)
+
+        encoder_layers = TransformerEncoderLayer(
+            embed_dim, n_heads, dim_feedforward=d_ffn, dropout=dropout, batch_first=True
+        )
+        self.transformer_encoder = TransformerEncoder(encoder_layers, n_layers)
+
+    def forward(self, state, img_feats):
+        """
+        Args:
+            state: torch.Tensor, shape [batch_size, n_obs_steps, state_dim]
+            img_feats: torch.Tensor, shape [batch_size, n_obs_steps, n_patches, embed_dim]
+        Returns:
+            readout_embeds: torch.Tensor, shape [batch_size, n_obs_steps, n_readouts, embed_dim]
+        """
+        b, t, *_ = img_feats.size()
+        img_proj = self.img_proj(img_feats)
+        state_proj = self.state_proj(state)
+        x = torch.cat(
+            [
+                img_proj,
+                state_proj.view((b, t, 1, -1)),
+                repeat(self.readout_tokens, "1 t r d -> b t r d", b=b),
+            ],
+            dim=2,
+        )
+        x = rearrange(x, "b t l f -> b (t l) f")
+        if self.add_positional_encoding:
+            x = self.pos_emb(x)
+        x = self.transformer_encoder(x, mask=self.blockwise_causal_mask)
+        x = rearrange(x, "b (t l) f -> b t l f", t=t)
+        readout_embeds = x[:, :, -self.n_readouts :, :]
+        return readout_embeds
+
+
+class DiffusionActionHead(nn.Module):
+    """Diffusion Action Head for Octo, as described above.
+
+    Args:
+        time_dim: int, dimension of the denoising iteration projection.
+        cond_dim: int, dimension of the conditioning features.
+        actions_dim: int, dimension of the output actions (should be pred_horizon * action_dim).
+        n_diffusion_head_layers: int, number of diffusion head layers.
+        diffusion_head_dim: int, dimension of the hidden layers in the diffusion head.
+    """
+
+    def __init__(self, time_dim, cond_dim, actions_dim, n_diffusion_head_layers, diffusion_head_dim):
+        super().__init__()
+
+        self.ff = FourierFeatures(time_dim)
+        self.time_ff_encoder = MLP(time_dim, (2 * time_dim, time_dim))
+        self.net = MLPResNet(
+            time_dim + cond_dim + actions_dim,
+            actions_dim,
+            hidden_dim=diffusion_head_dim,
+            num_layers=n_diffusion_head_layers,
+        )
+
+    def forward(self, readout_embeds, time, actions):
+        """
+        Args:
+            readout_embeds: torch.Tensor, shape [batch_size, n_obs_steps, n_readouts, embed_dim]
+            time: torch.Tensor, shape [batch_size, 1]
+            actions: torch.Tensor, shape [batch_size, pred_horizon * action_dim]
+        Returns:
+            eps_pred: torch.Tensor, shape [batch_size, pred_horizon * action_dim]
+        """
+        mean_readouts_embed = readout_embeds.mean(dim=(1, 2))
```

Comment:
i'm unsure if there are any pitfalls to doing this, if anyone has opinions. the original Octo uses mean readout embeddings of the same obs timestep (and generates multiple trajectories). another option here maybe to just use the last readout embedding?